### DISCUSSION TITLE
Non-pickleable outputs

### DISCUSSION THREAD

**Discussion by bra-fsn** [pipefunc/pipefunc#325] [General]
Is it possible to return non-pickleable outputs?
Currently doing this fails (on a parallel map) with:
```
    res = pipeline.map(
          ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pipefunc/_pipeline.py", line 680, in map
    return run(
           ^^^^
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_run.py", line 125, in run
    _run_and_process_generation(
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_run.py", line 587, in _run_and_process_generation
    _outputs = _process_task(func, tasks[func], run_info.run_folder, store, executor)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_run.py", line 619, in _process_task
    outputs_list = list(r)
                   ^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/_base.py", line 619, in result_iterator
    yield _result_or_cancel(fs.pop())
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/_base.py", line 317, in _result_or_cancel
    return fut.result(timeout)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_run.py", line 338, in _run_iteration_and_process
    _update_file_array(func, file_arrays, shape, shape_mask, index, outputs)
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_run.py", line 354, in _update_file_array
    file_array.dump(output_key, _output)
  File "/usr/local/lib/python3.12/site-packages/pipefunc/map/_filearray.py", line 258, in dump
    dump(value, self._key_to_file(key))  # type: ignore[arg-type]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/pipefunc/_utils.py", line 41, in dump
    cloudpickle.dump(obj, f)
  File "/usr/local/lib/python3.12/site-packages/cloudpickle/cloudpickle.py", line 1461, in dump
    Pickler(file, protocol=protocol, buffer_callback=buffer_callback).dump(obj)
  File "/usr/local/lib/python3.12/site-packages/cloudpickle/cloudpickle.py", line 1245, in dump
    return super().dump(obj)
           ^^^^^^^^^^^^^^^^^
TypeError: cannot pickle 'MyOwnClass' object
```

I've already tried to pass `cache=False` to all `pipefunc` decorators and `Pipeline(funcs, cache_type="simple")`

**Comment by basnijholt** [comment]
AFK now, will add more details later. 

With `parallel=True` I don't think there is a way around it. Without parallel you could set `storage="dict"` in `map(...)`. 

EDIT ([from here](https://github.com/pipefunc/pipefunc/discussions/325#discussioncomment-10677906)):

@bra-fsn I just released [0.32.0](https://github.com/pipefunc/pipefunc/releases/tag/v0.32.0).

This should now "just work":
```python
from concurrent.futures import ThreadPoolExecutor
from pipefunc.map import DictArray, register_storage
from pipefunc import Pipeline, pipefunc

class Unpicklable:
    def __getstate__(self):
        msg = "Unpicklable object"
        raise RuntimeError(msg)

@pipefunc(output_name="y")
def f():
    return Unpicklable()


@pipefunc(output_name="z")
def g():
    return 1

def h(y, z):
    return y, z

pipeline = Pipeline([f, g, h])

ex = ThreadPoolExecutor()
pipeline.map({}, storage="dict", executor=ex)
```

**Reply by basnijholt** [reply]
Oh it turns out that even with `dict` it tries to persist the result to disk.

I will make this optional.

Can you really not make the object serializable? In most cases this is pretty easy. Just defining a `__getstate__` and `__setstate__`.

**Reply by bra-fsn** [reply]
> Oh it turns out that even with `dict` it tries to persist the result to disk.
> 
> I will make this optional.
> 
> Can you really not make the object serializable? In most cases this is pretty easy. Just defining a `__getstate__` and `__setstate__`.

As it's kept in memory and I use the results from there, I really see no point in persisting everything, so optionally leaving this out would be great.
It's a Rust binding and its state. Sure, it could be solved, but it would be better to skip the overhead of serialization/storage altogether.
BTW, I did a quick check on the docs, but couldn't find the reason/value in this. Is this related to the default `ProcessPoolExecutor` (values are passed serialized)?
I use `ThreadPoolExecutor`, so if that's the case, I don't need it.
So making this optional would be great, thank you!

**Reply by basnijholt** [reply]
@bra-fsn, I am working on a run mode that does not write *anything* to disk here: https://github.com/pipefunc/pipefunc/pull/327

**Reply by basnijholt** [reply]
@bra-fsn I just released [0.32.0](https://github.com/pipefunc/pipefunc/releases/tag/v0.32.0).

This should now "just work":
```python
from concurrent.futures import ThreadPoolExecutor
from pipefunc.map import DictArray, register_storage
from pipefunc import Pipeline, pipefunc

class Unpicklable:
    def __getstate__(self):
        msg = "Unpicklable object"
        raise RuntimeError(msg)

@pipefunc(output_name="y")
def f():
    return Unpicklable()


@pipefunc(output_name="z")
def g():
    return 1

def h(y, z):
    return y, z

pipeline = Pipeline([f, g, h])

ex = ThreadPoolExecutor()
pipeline.map({}, storage="dict", executor=ex)
```

**Reply by bra-fsn** [reply]
> @bra-fsn I just released [0.32.0](https://github.com/pipefunc/pipefunc/releases/tag/v0.32.0).
> 
> This should now "just work":

🙇‍♂️
This is cool, thank you! I've noticed that this doesn't yet work:
https://github.com/pipefunc/pipefunc/pull/327#issuecomment-2357643505

**Reply by basnijholt** [reply]
I just released [0.32.1](https://github.com/pipefunc/pipefunc/releases/tag/v0.32.1) with the fix! :tada:

Let me know if you are experiencing any other issues! 😄

**Reply by bra-fsn** [reply]
> I just released [0.32.1](https://github.com/pipefunc/pipefunc/releases/tag/v0.32.1) with the fix! 🎉
> 
> Let me know if you are experiencing any other issues! 😄

Resource usage with the old and the new (`storage="dict"`) approach:

|               | file writes | storage=host |
|---------------|-------------|--------------|
| Usr time [s]      | 5.95        | 5.01         |
| Sys time [s]      | 2.29        | 1.96         |
| Algo runtime [s] | 2.15        | 1.11         |
| Max RSS [kiB]      | 430976      | 423464       |
| Total runtime [s] | 3.07     | 1.97      |

Yay, thanks!

**Reply by basnijholt** [reply]
Awesome! Although I am not quite sure what I am looking at 😄 

Could you share a little bit about your use-case? Just curious!

**Reply by bra-fsn** [reply]
> Awesome! Although I am not quite sure what I am looking at 😄
> 
> Could you share a little bit about your use-case? Just curious!

I'm just playing with it now, the above use case loads data from network (for the measurement it was loading from local disk), processes it and does some transformation/network API calls on those. So the processing for each chunk of data is CPU-bound, but all other tasks are mostly IO(network)-bound, so havig a thread based approach works quite well.
Generally what I wanted to have is something, which is flexible for implementing the whole workflow in a DAG-based approach with the ability to easily extend it, which pipefunc fulfills pretty neatly.

I've written some code previously in pytest (nothing test related) because of its fixtures, which remotely resembles the DAG approach, but it doesn't really support good concurrency and dependency handling.
I'm glad that you've written pipefunc and I've found it, it will make my life a lot easier in several cases!

**Reply by basnijholt** [reply]
Interesting!

I have been toying with the idea to be able to provide different executors for different pipeline functions.
In your case this might make sense, e.g., using a `ThreadPoolExecutor` for the IO-bound stuff and a `ProcessPoolExecutor` for the CPU-bound stuff. Then some trivial steps could run in the current kernel too for those pipeline functions that take a µs.

Thanks a lot for your inputs!

**Reply by basnijholt** [reply]
BTW https://pipefunc.readthedocs.io/en/latest/faq/#mixing-executors-and-storage-backends-for-i-o-bound-and-cpu-bound-work
